File: leet/array/lengthOfLongestSubstring.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        unordered_map<char, int> m;
        int i=0, maxlength=0;        
        
        for (int j=0; j<s.length(); j++){
            
            // if found
            if (m.find(s[j]) != m.end()){
                i = max(m[s[j]], i);
            } 
                            
            maxlength = max( (j - i + 1), maxlength);

            // Store result
            m[s[j]] = j+1;
        }
        
        return maxlength;
    }
};

File: leet/array/lengthOfLongestSubstring.py
"""
 tmmzuxt
  i
       j
               
"""

class Solution:
    
    def lengthOfLongestSubstring(self, s: str) -> int:
        maxlength=0
        i = 0
        used = {}
        
        for j,c in enumerate(s):
            if c in used and i <= used[c]:
                # move i to after found char
                i = max(used[c]+1, i)
            else:
                maxlength = max(maxlength, j-i+1,)
                
            used[c] = j

        return maxlength